sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/CertificateValidationCheckBase.cs

```diff
@@ -0,0 +1,298 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2019 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class CertificateValidationCheckBase<TMethodSyntax, TArgumentSyntax, TExpressionSyntax, TIdentifierNameSyntax, TAssignmentExpressionSyntax, TInvocationExpressionSyntax, TParameterSyntax, TVariableSyntax, TLambdaSyntax> : SonarDiagnosticAnalyzer
+        where TMethodSyntax : SyntaxNode
+        where TArgumentSyntax : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+        where TIdentifierNameSyntax : SyntaxNode
+        where TAssignmentExpressionSyntax : SyntaxNode
+        where TInvocationExpressionSyntax : SyntaxNode
+        where TParameterSyntax : SyntaxNode
+        where TVariableSyntax : SyntaxNode
+        where TLambdaSyntax : SyntaxNode
```

Comment:
There are actually only 2 methods relying all these generic types to be specifc. I suggest to have a single generic `TSyntaxNode` with `where TSyntaxNode : SyntaxNode`, and override the 2 methods left that rely on specific types: `ArgumentLocations` and `IdentifierLocations`, with the logic inside `ArgumentLocations` refactored into helper methods.
WDYT ?

Reply:
I would be fine with `IdentifierLocations`. But `ArgumentLocations` contains the core of the algorithm and contains heavy logic. I would rather not to couple this piece of code. When having this and `IdentifierLocations`, it creates two usage places to justify this design.

I consider this a common problem when writing common C# and VB.NET rule. I plan to address in the future update of C# vs. VB.NET rule unification.